src/analyser.cpp

```diff
@@ -0,0 +1,1509 @@
+
+    void setVariable(const VariablePtr &variable,
+                     bool checkInitialValue = true);
+
+    void makeVoi();
+    void makeState();
+};
+
+using AnalyserInternalVariablePtr = std::shared_ptr<AnalyserInternalVariable>;
+
+AnalyserInternalVariable::AnalyserInternalVariable(const VariablePtr &variable)
+{
+    setVariable(variable);
+}
+
+void AnalyserInternalVariable::setVariable(const VariablePtr &variable,
+                                           bool checkInitialValue)
+{
+    if (checkInitialValue && !variable->initialValue().empty()) {
+        // The variable has an initial value, so it can either be a constant or
+        // a state. By default, we consider it to be a constant and, if we find
+        // an ODE for that variable, we will know that it was actually a state.
+
+        mType = Type::CONSTANT;
+
+        mInitialisingVariable = variable;
+    }
+
+    mVariable = variable;
+}
+
+void AnalyserInternalVariable::makeVoi()
+{
+    mType = Type::VARIABLE_OF_INTEGRATION;
+}
+
+void AnalyserInternalVariable::makeState()
+{
+    if (mType == Type::UNKNOWN) {
+        mType = Type::SHOULD_BE_STATE;
+    } else if (mType == Type::CONSTANT) {
+        mType = Type::STATE;
+    }
+}
+
+#ifdef SWIG
+struct AnalyserEquation
+#else
+struct AnalyserInternalEquation: public std::enable_shared_from_this<AnalyserInternalEquation>
+#endif
+{
+    enum struct Type
+    {
+        UNKNOWN,
+        TRUE_CONSTANT,
+        VARIABLE_BASED_CONSTANT,
+        RATE,
+        ALGEBRAIC
+    };
+
+    size_t mOrder = MAX_SIZE_T;
+    Type mType = Type::UNKNOWN;
+
+    std::vector<AnalyserInternalEquationPtr> mDependencies;
+
+    AnalyserEquationAstPtr mAst;
+
+    std::vector<AnalyserInternalVariablePtr> mVariables;
+    std::vector<AnalyserInternalVariablePtr> mOdeVariables;
+
+    AnalyserInternalVariablePtr mVariable;
+    ComponentPtr mComponent = nullptr;
+
+    bool mComputedTrueConstant = true;
+    bool mComputedVariableBasedConstant = true;
+
+    bool mIsStateRateBased = false;
+
+    explicit AnalyserInternalEquation(const ComponentPtr &component);
+
+    void addVariable(const AnalyserInternalVariablePtr &variable);
+    void addOdeVariable(const AnalyserInternalVariablePtr &odeVariable);
+
+    static bool containsNonUnknownVariables(const std::vector<AnalyserInternalVariablePtr> &variables);
+    static bool containsNonConstantVariables(const std::vector<AnalyserInternalVariablePtr> &variables);
+
+    static bool knownVariable(const AnalyserInternalVariablePtr &variable);
+    static bool knownOdeVariable(const AnalyserInternalVariablePtr &odeVariable);
+
+    bool check(size_t & equationOrder, size_t & stateIndex, size_t & variableIndex);
+};
+
+AnalyserInternalEquation::AnalyserInternalEquation(const ComponentPtr &component)
+    : mAst(AnalyserEquationAst::create())
+    , mComponent(component)
+{
+}
+
+void AnalyserInternalEquation::addVariable(const AnalyserInternalVariablePtr &variable)
```

Comment:
All "add" functions should return a bool as to whether or not it's happened.  See PR #621

Reply:
You are again comparing internal methods with API methods. Internally, we know (or should know!) what we are doing and, in this particular case, there is no need to return a boolean.